File: frontend/src/api/jobs.ts
import client from './client'
import type { Job, JobListResponse } from '../types'

export const getJobs = async (page = 1, pageSize = 20): Promise<JobListResponse> => {
  const { data } = await client.get<JobListResponse>('/jobs/', { params: { page, page_size: pageSize } })
  return data
}

export const getActiveJobs = async (): Promise<Job[]> => {
  const { data } = await client.get<Job[]>('/jobs/active')
  return data
}

export const getJob = async (id: number): Promise<Job> => {
  const { data } = await client.get<Job>(`/jobs/${id}`)
  return data
}

export const createJob = async (payload: {
  title: string
  description: string
  department?: string
  required_skills: string[]
  preferred_skills: string[]
  min_experience_years: number
  max_experience_years?: number
  status?: string
}): Promise<Job> => {
  const { data } = await client.post<Job>('/jobs/', payload)
  return data
}

export const updateJob = async (id: number, payload: Partial<Job>): Promise<Job> => {
  const { data } = await client.put<Job>(`/jobs/${id}`, payload)
  return data
}

export const deleteJob = async (id: number): Promise<void> => {
  await client.delete(`/jobs/${id}`)
}

File: app/api/v1/jobs.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_current_admin, get_current_user
from app.core.exceptions import NotFoundException
from app.database import get_db
from app.models.user import User
from app.schemas.job import JobCreate, JobListResponse, JobResponse, JobUpdate
from app.services.job_service import JobService

router = APIRouter(prefix="/jobs", tags=["Jobs"])


@router.post("/", response_model=JobResponse, status_code=201, summary="Create a job posting")
async def create_job(
    data: JobCreate, db: AsyncSession = Depends(get_db), _: User = Depends(get_current_user)
):
    return await JobService(db).create(data)


@router.get("/", response_model=JobListResponse, summary="List all job postings")
async def list_jobs(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    _: User = Depends(get_current_user),
):
    jobs, total = await JobService(db).get_all(page=page, page_size=page_size)
    return JobListResponse(total=total, page=page, page_size=page_size, jobs=jobs)


@router.get("/active", response_model=list[JobResponse], summary="List active job postings")
async def get_active_jobs(db: AsyncSession = Depends(get_db), _: User = Depends(get_current_user)):
    return await JobService(db).get_active()


@router.get("/{job_id}", response_model=JobResponse, summary="Get a job by ID")
async def get_job(
    job_id: int, db: AsyncSession = Depends(get_db), _: User = Depends(get_current_user)
):
    job = await JobService(db).get_by_id(job_id)
    if not job:
        raise HTTPException(status_code=404, detail=f"Job {job_id} not found")
    return job


@router.put("/{job_id}", response_model=JobResponse, summary="Update a job posting")
async def update_job(
    job_id: int,
    data: JobUpdate,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(get_current_user),
):
    try:
        return await JobService(db).update(job_id, data)
    except NotFoundException as exc:
        raise HTTPException(status_code=404, detail=exc.message)


@router.delete("/{job_id}", status_code=204, summary="Delete a job posting (admin only)")
async def delete_job(
    job_id: int, db: AsyncSession = Depends(get_db), _: User = Depends(get_current_admin)
):
    deleted = await JobService(db).delete(job_id)
    if not deleted:
        raise HTTPException(status_code=404, detail=f"Job {job_id} not found")
